### PR TITLE
Issue 4332

### PR DESCRIPTION
**Pull request by rahuldino**
connects #4332 

Added new validation for the fan assessment, field data section and removed other fields.

### CODE REVIEW

File: src/app/fsat/fan-field-data/fan-field-data.component.html [hunk 1]
```diff
@@ -83,19 +86,22 @@ <h3>Field Data</h3>
       <div class="input-group">
         <input name="outletPressure" type="number" min="0.1" step="any" class="form-control"
           formControlName="outletPressure" id="{{'outletPressure_'+idString}}" (input)="save()"
-          (focus)="focusField('outletPressure')" onfocus="this.select();" [readonly]="!selected">
+          (focus)="focusField('outletPressure')" onfocus="this.select();" [readonly]="!selected" (input)="compareValues()">
```
Comment (rmroot) [inline src/app/fsat/fan-field-data/fan-field-data.component.html:89]: Can't have two bindings to (input) event on the input element

File: src/app/fsat/fan-field-data/fan-field-data.component.html [hunk 2]
```diff
@@ -83,19 +86,22 @@ <h3>Field Data</h3>
...
         *ngIf="fieldDataForm.controls.outletPressure.invalid && !fieldDataForm.controls.outletPressure.pristine">
         <span *ngIf="fieldDataForm.controls.outletPressure.errors.required">Value Required</span>
-        <span *ngIf="fieldDataForm.controls.outletPressure.errors.min">Value can't be less than
-          {{fieldDataForm.controls.outletPressure.errors.min.min}} <span
-            [innerHTML]="settings.fanPressureMeasurement | settingsLabel"></span></span>
+        <div *ngIf="compareError">
+        <span>Outlet Pressure cannot be less than Inlet Pressure</span>    
+      </div>
+      </span>
+        <span class="alert-warning pull-right small"
```
Comment (rmroot) [inline src/app/fsat/fan-field-data/fan-field-data.component.html:99]: alert-warning should be used with a warning and not invalid

File: src/app/fsat/fan-field-data/fan-field-data.component.html [hunk 3]
```diff
@@ -83,19 +86,22 @@ <h3>Field Data</h3>
...
+          (focus)="focusField('outletPressure')" onfocus="this.select();" [readonly]="!selected" (input)="compareValues()">
         <span class="input-group-addon units" [innerHTML]="settings.fanPressureMeasurement | settingsLabel"></span>
       </div>
       <span class="alert-danger pull-right small"
         *ngIf="fieldDataForm.controls.outletPressure.invalid && !fieldDataForm.controls.outletPressure.pristine">
         <span *ngIf="fieldDataForm.controls.outletPressure.errors.required">Value Required</span>
-        <span *ngIf="fieldDataForm.controls.outletPressure.errors.min">Value can't be less than
-          {{fieldDataForm.controls.outletPressure.errors.min.min}} <span
-            [innerHTML]="settings.fanPressureMeasurement | settingsLabel"></span></span>
+        <div *ngIf="compareError">
```
Comment (rmroot) [inline src/app/fsat/fan-field-data/fan-field-data.component.html:95]: this should be a part of the form validation. Should be an error on outletPressure

File: src/app/fsat/fan-field-data/fan-field-data.component.ts [hunk 4]
```diff
@@ -433,4 +435,12 @@ export class FanFieldDataComponent implements OnInit {
   //     return false;
   //   }
   // }
+  
+  compareValues() {
```
Comment (rmroot) [inline src/app/fsat/fan-field-data/fan-field-data.component.ts:439]: This is an error on the form and should mark the form invalid.

File: src/app/fsat/fan-field-data/fan-field-data.service.ts [hunk 5]
```diff
@@ -15,7 +15,7 @@ export class FanFieldDataService {
     }
     let form: FormGroup = this.formBuilder.group({
       operatingHours: [obj.operatingHours, [Validators.required, Validators.min(0), Validators.max(8760)]],
-      flowRate: [obj.flowRate, [Validators.required, Validators.min(0)]],
+      flowRate: [obj.flowRate, [Validators.required, Validators.min(0.1)]],
```
Comment (rmroot) [inline src/app/fsat/fan-field-data/fan-field-data.service.ts:18]: Use GreaterThanValidator

GreaterThanValidator.greaterThan(0)

File: src/app/fsat/fsat-fluid/fsat-fluid.component.html [hunk 6]
```diff
@@ -153,7 +153,8 @@ <h3>Fluid</h3>
 
     </div>
 
-    <div class="form-group" *ngIf="gasDensityForm.controls.gasType.value == 'OTHER'"
+    <div class="form-group"
+      *ngIf="!gasDensityForm.controls.gasType.value == 'OTHER' || !gasDensityForm.controls.gasType.value == 'AIR'"
```
Comment (rmroot) [inline src/app/fsat/fsat-fluid/fsat-fluid.component.html:157]: should be

`*ngIf="gasDensityForm.controls.gasType.value != 'OTHER'"`

File: src/app/fsat/fsat.service.ts [hunk 7]
```diff
@@ -9,7 +9,6 @@ import { Settings } from '../shared/models/settings';
 import { ConvertFsatService } from './convert-fsat.service';
 import { ConvertUnitsService } from '../shared/convert-units/convert-units.service';
 import { FanEfficiencyInputs } from '../calculator/fans/fan-efficiency/fan-efficiency.service';
-import { ConvertFanAnalysisService } from '../calculator/fans/fan-analysis/convert-fan-analysis.service';
```
Comment (rmroot) [inline src/app/fsat/fsat.service.ts:12]: We don't want to remove all the conversion logic...

File: src/app/fsat/fsat.service.ts [hunk 8]
```diff
@@ -48,36 +47,27 @@ export class FsatService {
 
   fan203(input: Fan203Inputs, settings: Settings): Fan203Results {
     let inputCpy: Fan203Inputs = JSON.parse(JSON.stringify(input));
-    inputCpy = this.convertFanAnalysisService.convertFan203DataForCalculations(inputCpy, settings);
```
Comment (rmroot) [inline src/app/fsat/fsat.service.ts:51]: Still need all this conversion logic in all of these places...

**Review by rmroot — CHANGES_REQUESTED** [review]
1- The greater than and less than warnings checks for inlet pressure and outlet pressure should be a part of the checkFieldDataWarnings call and FanFieldDataWarnings object.


2- Since OutletPressure > InletPressure is an ERROR it should be apart of the form. 

In fan-field-data.service.ts getFormFromObj():

outletPressure: [obj.outletPressure, [Validators.required, Validators.min(obj.inletPressure)]`

IMPORTANT: When the inletPressure in the form changes you will need to update the validators on the form in the component. Ask me if you need help doing this.

File: src/app/fsat/fsat-fluid/fsat-fluid.component.html [hunk 9]
```diff
@@ -153,7 +153,8 @@ <h3>Fluid</h3>
 
     </div>
 
-    <div class="form-group" *ngIf="gasDensityForm.controls.gasType.value == 'OTHER'"
+    <div class="form-group"
+    *ngIf="gasDensityForm.controls.gasType.value != 'OTHER' && gasDensityForm.controls.gasType.value != 'AIR'"
       [ngClass]="{'indicate-different': isSpecificGravityDifferent(), 'invalid':gasDensityForm.controls.specificGravity.invalid}">
```
Comment (rmroot) [inline src/app/fsat/fsat-fluid/fsat-fluid.component.html:158]: Shouldn't have the check for AIR. We want to display this field if AIR is selected

**Review by rmroot — CHANGES_REQUESTED** [review]

**Review by nbintertech — APPROVED** [review]

**Review by rmroot — APPROVED** [review]

**Pull request merged**